test/es-module/test-esm-extension-lookup-deprecation.mjs

```diff
@@ -0,0 +1,121 @@
+import { spawnPromisified } from '../common/index.mjs';
+import * as tmpdir from '../common/tmpdir.js';
+
+import assert from 'node:assert';
+import { mkdir, writeFile } from 'node:fs/promises';
+import * as path from 'node:path';
+import { execPath } from 'node:process';
+import { describe, it, before } from 'node:test';
+
+describe('ESM in main field', { concurrency: true }, () => {
+  before(() => tmpdir.refresh());
+
+  it('should handle fully-specified relative path without any warning', async () => {
+    const cwd = path.join(tmpdir.path, Math.random().toString());
+    const pkgPath = path.join(cwd, './node_modules/pkg/');
+    await mkdir(pkgPath, { recursive: true });
+    await writeFile(path.join(pkgPath, './index.js'), 'console.log("Hello World!")');
+    await writeFile(path.join(pkgPath, './package.json'), JSON.stringify({
+      main: './index.js',
+      type: 'module',
+    }));
+    const { code, stdout, stderr } = await spawnPromisified(execPath, [
+      '--input-type=module',
+      '--eval', 'import "pkg"',
+    ], { cwd });
+
+    assert.strictEqual(stderr, '');
+    assert.match(stdout, /^Hello World!\r?\n$/);
+    assert.strictEqual(code, 0);
+  });
+  it('should handle fully-specified absolute path without any warning', async () => {
+    const cwd = path.join(tmpdir.path, Math.random().toString());
+    const pkgPath = path.join(cwd, './node_modules/pkg/');
+    await mkdir(pkgPath, { recursive: true });
+    await writeFile(path.join(pkgPath, './index.js'), 'console.log("Hello World!")');
+    await writeFile(path.join(pkgPath, './package.json'), JSON.stringify({
+      main: path.join(pkgPath, './index.js'),
+      type: 'module',
+    }));
+    const { code, stdout, stderr } = await spawnPromisified(execPath, [
+      '--input-type=module',
+      '--eval', 'import "pkg"',
+    ], { cwd });
+
+    assert.strictEqual(stderr, '');
+    assert.match(stdout, /^Hello World!\r?\n$/);
+    assert.strictEqual(code, 0);
+  });
```

Comment:
Node 20.3.0 emit deprecation warning for this case:

```
h4ad:node/ (fix-dep0151✗) $ node -v                                                                                                                                                                                                                                  
v20.3.0
h4ad:node/ (fix-dep0151✗) $ node --expose-internals --no-warnings ./test/es-module/test-esm-extension-lookup-deprecation.mjs                                                                                                                                        
▶ ESM in main field
  ✔ should handle fully-specified relative path without any warning (32.748089ms)
  ✖ should handle fully-specified absolute path without any warning (31.90193ms)
    AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
    + actual - expected
    
    + '(node:171907) [DEP0151] DeprecationWarning: Package /home/h4ad/Projects/opensource/node/test/.tmp.0/0.06212600443072347/node_modules/pkg/ has a "main" field set to "/home/h4ad/Projects/opensource/node/test/.tmp.0/0.06212600443072347/node_modules/pkg/index.js", excluding the full filename and extension to the resolved file at "index.js", imported from /home/h4ad/Projects/opensource/node/test/.tmp.0/0.06212600443072347/[eval1].\n' +
    +   ' Automatic extension resolution of the "main" field is deprecated for ES modules.\n' +
    +   '(Use `node --trace-deprecation ...` to show where the warning was created)\n'
    - ''
        at TestContext.<anonymous> (file:///home/h4ad/Projects/opensource/node/test/es-module/test-esm-extension-lookup-deprecation.mjs:45:12)
        at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
        at async Test.run (node:internal/test_runner/test:570:9)
        at async Promise.all (index 1)
        at async Suite.run (node:internal/test_runner/test:806:7)
        at async startSubtest (node:internal/test_runner/harness:204:3) {
      generatedMessage: true,
      code: 'ERR_ASSERTION',
      actual: '(Use `node --trace-deprecation ...` to show where the warning was created)\n...',
      expected: '',
      operator: 'strictEqual'
    }

  ✔ should emit warning when "main" and "exports" are missing (32.352502ms)
  ✔ should emit warning when "main" is falsy (32.283562ms)
  ✔ should emit warning when "main" is a relative path without extension (37.497414ms)
  ✔ should emit warning when "main" is an absolute path without extension (37.727778ms)
▶ ESM in main field (40.050949ms)

ℹ tests 6
ℹ suites 1
ℹ pass 5
ℹ fail 1
ℹ cancelled 0
ℹ skipped 0
ℹ todo 0
ℹ duration_ms 0.197433

✖ failing tests:

✖ should handle fully-specified absolute path without any warning (31.90193ms)
  AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
  + actual - expected
  
  + '(node:171907) [DEP0151] DeprecationWarning: Package /home/h4ad/Projects/opensource/node/test/.tmp.0/0.06212600443072347/node_modules/pkg/ has a "main" field set to "/home/h4ad/Projects/opensource/node/test/.tmp.0/0.06212600443072347/node_modules/pkg/index.js", excluding the full filename and extension to the resolved file at "index.js", imported from /home/h4ad/Projects/opensource/node/test/.tmp.0/0.06212600443072347/[eval1].\n' +
  +   ' Automatic extension resolution of the "main" field is deprecated for ES modules.\n' +
  +   '(Use `node --trace-deprecation ...` to show where the warning was created)\n'
  - ''
      at TestContext.<anonymous> (file:///home/h4ad/Projects/opensource/node/test/es-module/test-esm-extension-lookup-deprecation.mjs:45:12)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async Test.run (node:internal/test_runner/test:570:9)
      at async Promise.all (index 1)
      at async Suite.run (node:internal/test_runner/test:806:7)
      at async startSubtest (node:internal/test_runner/harness:204:3) {
    generatedMessage: true,
    code: 'ERR_ASSERTION',
    actual: '(Use `node --trace-deprecation ...` to show where the warning was created)\n...',
    expected: '',
    operator: 'strictEqual'
  }
```

Reply:
The error message is clearly wrong though, I don't think there's much value in keeping the old behavior – well maybe there's a perf incentive

Reply:
I did a test and don't have any perf gains in keeping the old behavior, makes sense to change this message if the old one was not accurate.

Reply:
Yep! Thanks for the fix